### ISSUE TITLE
TURN in the enterprise

### ISSUE THREAD

**Issue by venkiahn** [coturn/coturn#1347]
I am working on implementing a VR pixel streaming capability from an AWS VPC to on prem clients.  The experience is rendered through Unreal Engine and we're attempting to use STUN/TURN and Signalling however I can't find a pattern that supports having the STUN/TURN and Signalling EC2 instances within a private subnet have a public subnet NAT Gateway and Load Balancers for both these services.  Has anyone successfully implemented a pattern like this before?

**Comment by eakraly** [comment]
Hi @venkiahn 
I suppose you are trying to use pixel streaming from https://github.com/TensorWorks
They are amazing and can provide high level of service and support - maybe better than here. I worked with their team in the past so I am talking from experience.

Also, there is a recent post from AWS on exactly this topic which maybe relevant: https://aws.amazon.com/blogs/gametech/deploy-unreal-engines-pixel-streaming-at-scale-on-aws/

**Comment by venkiahn** [comment]
Hi @eakraly

Thank you for your response. I have read through TensorWorks documentation however these private STUN/TURN and Matchmaking services behind a Load Balancer and NAT GW is something they've obfuscated and would potentially require a paid service.  This recent post is amazing however it proposes having the STUN/TURN on the GPU instances which will need to reside in a public subnet with a public EIP.  Unfortunately this is not a pattern enterprises would readily accept.

**Comment by eakraly** [comment]
@venkiahn The goal of TURN server is to provide a connectivity when one or both sides are not "direct connection friendly". Trying to put TURN server "inside" of one of those "anti-connection friendly" sides defeats its purpose...

Alternatively, putting TURN on the edge of the VPC as a gateway can bridge outside and inside but it does not work with NAT gateway because TURN server acts as a gateway itself. TURN is not application level service but network level...

Maybe you can do double TURN connection where each side (both streaming and client) have connection through coturn on each side which acts as a gateway on each side so that you have streamer->turn->turn->client?

**Comment by venkiahn** [comment]
@eakraly I see your point about TURN being network level and not application level however there must be a way to load balance TURN and ensure its highly available?

**Comment by jonesmz** [comment]
DNS round-robin + a service healthchecker that understands protocol specific health.

If you have a UDP protocol load balancer, you could use that, but how are you planning to load balance your load balancer and ensure high availability of your load balancer?

**Comment by eakraly** [comment]
@venkiahn load balancing is a concept - it does not mean you have to have a "load balancing device" in front of the service to achieve it.

With coturn, as I described, the common patter is to use DNS round-robin as a load balancing implementation in front of a pool of coturn servers. Another approach is to use ALTERNATE-SERVER approach but I have not used it and IMHO it is worse than DNS

**Comment by venkiahn** [comment]
@eakraly Yes my customer wants to have an AWS Load Balancer (ALB or NLB) in front of TURN in order to prevent the EC2 instance from being directly accessible to the internet via a public IP address.

**Comment by jonesmz** [comment]
I strongly recommend against having a load balancer in front of the turn servers. TURN is a stateful protocol. Once you establish a session with a TURN server, you cannot switch to another until you close that session.

But this seems to be pretty far outside of the realm of what the coturn project can realistically help you with unless you're wanting advice on how to add functionality to coturn to make it easier to setup an AWS load balancer to load balance the turnserver.

You'll need the following two things that are currently lacking in coturn:

1. a health checking mechanism that the AWS tooling can understand. At least the elastic load balancing product from AWS only speaks TCP / HTTP to understand instance health, and coturn does not currently afford anything related to that. I'm working on creating a standalone program that can be triggered by the ELB health check via a systemd socket unit, but have had very little time to take that forward recently.
2. a graceful shutdown mechanism (e.g. let all existing sessions be ended by the client devices naturally). My understanding is that coturn does not have any functionality for this built in. It can be implemented largely outside of the coturn codebase, but it'd be nice if coturn itself understood the idea of "Stop accepting new allocations, let all the current allocations continue, and exit when there are zero allocations".


And since I don't, myself, have any specific knowledge of how ALB or NLB work (I have only looked at ELB), i'm not sure if there's anything specific needed for those, but if there are things specific to ALB and/or NLB load balancers that coturn should be able to do, those can be submitted to the coturn codebase as reference/examples.

**Comment by rssissoko600** [comment]
@venkiahn Were you able to move forward with the need from your client? Im in a similar restraint: 

> @eakraly Yes my customer wants to have an AWS Load Balancer (ALB or NLB) in front of TURN in order to prevent the EC2 instance from being directly accessible to the internet via a public IP address.

**Comment by venkiahn** [comment]
@rssissoko600 yes I was able to use an autoscaling group behind an NLB, configured CoTurn to use the NLB's public IP address and the respective server's private IP address with good success.

**Comment by eakraly** [comment]
> @rssissoko600 yes I was able to use an autoscaling group behind an NLB, configured CoTurn to use the NLB's public IP address and the respective server's private IP address with good success.

@venkiahn If you could expand on your solution with more details it would be great! Make a post somewhere - even better. I can add your explanation to the docs if you agree.

**Comment by rssissoko600** [comment]
@eakraly and @venkiahn I agree this will be extremely helpful to those working behind private restrictive networks. I have been working on trying discover the limitations coturn has with private networks for months now

**Comment by kochanowski** [comment]
hey @venkiahn any chance you can share your solution running coturn behind NLB?
